pkg/controller/jobframework/reconciler.go

```diff
@@ -221,6 +245,10 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 		if err != nil {
 			log.Error(err, "Updating workload status")
 		}
+		// Execute job finalization logic
+		if err := r.finalizeJob(ctx, job); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
we should also make sure to call `finalizeJob` even if `wl == nil`.

Reply:
Please take a look at https://github.com/kubernetes-sigs/kueue/pull/1072/commits/7a61e8d00ddf443d044bea9fae252ee61002ff52#diff-552ab657667de70900b502e2617bcb405fbf4203ff08b2a35d7c87f2d637a75bR431

Reply:
🤔 I don't think that's right.
We cannot finalize a generic job when calling Stop because, in general, Stop doesn't imply termination.

In the case of Pod, Stop causes termination, but also we will get a separate reconciliation once the Pod finishes.

So I think the correct solution is something along these lines:

```golang
if job.finished() {
  if workload != nil {
     workload.UpdateStatus()
  }
  r.finalizeJob(job)
}
```

Don't forget to add a test case for it.

Reply:
Added a check in `ensureOneWorkload` function:

https://github.com/kubernetes-sigs/kueue/blob/1c7a932045462bf96865a33213d228549dd02634/pkg/controller/jobframework/reconciler.go#L431-L435